### PR TITLE
Build the docker-engine RPM for Amazon Linux

### PR DESCRIPTION
**Pull request by iliana**
**- What I did**
Added `amazonlinux-latest` to the set of RPMs to be built.

This replaces #17283 which was rejected while we waited for an Amazon Linux container image. There is now such a thing. :)

Amazon Linux is a rolling release and `latest` will be more correct than building RPMs for each successive release version bump we make. By default, the Amazon Linux AMI points to "latest" instead of a releasever like "2016.09".

I'd like for this to be considered for cherry-picking onto the 1.13 release branch.

**- How I did it**
Modified `contrib/builder/rpm/amd64/generate.sh` to create an appropriate file for the `amazonlinux` image, then ran `generate.sh amazonlinux-latest`.

**- How to verify it**
`make rpm`

**- Description for the changelog**
Add RPM builder for Amazon Linux

**- A picture of a cute animal (not mandatory but encouraged)**
kitten

### CODE REVIEW

File: contrib/builder/rpm/amd64/amazonlinux-latest/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,18 @@
+#
+# THIS FILE IS AUTOGENERATED; SEE "contrib/builder/rpm/amd64/generate.sh"!
+#
+
+FROM amazonlinux:latest
+
+RUN yum groupinstall -y "Development Tools"
+RUN yum install -y btrfs-progs-devel device-mapper-devel glibc-static  libselinux-devel libtool-ltdl-devel pkgconfig selinux-policy selinux-policy-devel sqlite-devel  tar git cmake vim-common
```
Comment (justincormack) [inline contrib/builder/rpm/amd64/amazonlinux-latest/Dockerfile]: this line could be broken up a bit with some `\` (and the spacing is uneven)

File: contrib/builder/rpm/amd64/generate.sh [hunk 2]
```diff
@@ -111,15 +111,15 @@ for version in "${versions[@]}"; do
...
+		oraclelinux:6|amazonlinux:latest)
 			# doesn't use systemd, doesn't have a devel package for it
 			packages=( "${packages[@]/systemd-devel}" )
 			;;
 	esac
 
 	# opensuse & oraclelinx:6 do not have the right libseccomp libs
 	case "$from" in
-		opensuse:*|oraclelinux:6)
+		opensuse:*|oraclelinux:6|amazonlinux:latest)
```
Comment (justincormack) [inline contrib/builder/rpm/amd64/generate.sh]: As far as I can see Amazon Linux does not ship `libseccomp` at all. I don't think disabling seccomp is acceptable for distributing Docker; we understand that some of the older distros have incompatible old versions but it is a highly recommended part of Docker security, and I am not happy about adding a new distro that does not support seccomp.

**Review by justincormack — CHANGES_REQUESTED** [review]

File: contrib/builder/rpm/amd64/amazonlinux-latest/Dockerfile (hunk 1 above)
Comment (ianweller) [inline contrib/builder/rpm/amd64/amazonlinux-latest/Dockerfile]: This is autogenerated. I could fix the script but that doesn't seem within scope for this pull request, IMO.

File: contrib/builder/rpm/amd64/generate.sh (hunk 2 above)
Comment (ianweller) [inline contrib/builder/rpm/amd64/generate.sh]: libseccomp is something we're looking at bringing into Amazon Linux -- would it be acceptable still add this with the understanding that we can enable libseccomp in the near future?

File: contrib/builder/rpm/amd64/amazonlinux-latest/Dockerfile (hunk 1 above)
Comment (justincormack) [inline contrib/builder/rpm/amd64/amazonlinux-latest/Dockerfile]: ok

File: contrib/builder/rpm/amd64/generate.sh (hunk 2 above)
Comment (justincormack) [inline contrib/builder/rpm/amd64/generate.sh]: Not really, no. You could remove it so it was ready to build once you ship the library I guess.

**Comment by ecliptik** [comment]
While researching why our Amazon Linux 2016.09 production Docker hosts were running out of disk space, we discovered that we were running into the out-of-inodes issue with using the storage driver `overlay` described in issue https://github.com/docker/docker/issues/10613.

According to that issue, the solution is to use `overlay2`, but it's only available in Docker 1.12, which is currently not shipped with Amazon Linux. I came upon this issue while seeing if Docker 1.12 would come with Amazon Linux 2017.03.

As you can see this puts us in a Catch-22, and getting the inode issue resolved, and having Docker 1.12 is available in Amazon Linux 2017.03 (or sooner) is highly desirable for us.

**Comment by ecliptik** [comment]
@ianweller would it be okay to build the docker-engine 1.12 RPM using your PR locally and installing in on Amazon Linux 2016.09 as a temporary workaround until it's officially released? Or do you forsee issues with kernel or library incompatibilities?

**Comment by ianweller** [comment]
@ecliptik I mean, it's meant to run on Amazon Linux 2016.09, so :) I've not done anything beyond a very quick spot check that it runs a container with mostly default arguments.

@justincormack I've put libseccomp on our roadmap for (hopefully) sometime this month. Want me to leave this open with a [WIP] tag, or close it until we've got seccomp out the door?

**Comment by ecliptik** [comment]
@ianweller thank you for the reply, and I built the 1.12 RPM using this PR and tested it out on a Rancher host with Amazon Linux 2016.09 running ~40 containers and it seemed to do well enough.

I'm going to burn it in a bit more and will re-deploy an entire cluster for additional testing. If I notice any issues I'll report them.

**Comment by justincormack** [comment]
@ianweller happy to leave it open, whatever works for you.

**Comment by LK4D4** [comment]
@ianweller any news here?

**Comment by ianweller** [comment]
libseccomp is in our repos now but I haven't had a chance to update this pull request.

**Comment by callumj** [comment]
Docker 1.12 is now available via the package repo for Amazon Linux: https://forums.aws.amazon.com/ann.jspa?annID=4343

**Comment by CpuID** [comment]
+1 to this can't wait to see it finalised :)

**Comment by ilianaw** [comment]
Sorry for the delay on this — we have libseccomp in our repos now and I've fixed the relevant parts of `generate.sh`, as well as committing the new autogenerated `amazonlinux-latest/Dockerfile`.

**Comment by vieux** [comment]
ping @andrewhsu WDYT ?

**Comment by andrewhsu** [comment]
Technically looks like on the right track. I have to see how this is will fit in with our upcoming releases.

**Comment by vieux** [comment]
LGTM, any comment @justincormack ?

**Review by justincormack — APPROVED** [review]

**Comment by justincormack** [comment]
LGTM

**Comment by justincormack** [comment]
@ecliptik you can probably only use `overlay2` with this if you disable SELinux, as there are still some unresolved issues with the combination (probably, unless they are all fixed in the Amazon kernel). @ilianaw have you tested different storage drivers?, we should perhaps add some docs about storage driver recommendations.

**Comment by ilianaw** [comment]
@justincormack I haven't but some folks on the ECS team have, I think. I'll check tomorrow sometime.

**Comment by vieux** [comment]
LGTM, let's address the remaining comments, if any in a separate PR.

**Comment by ilianaw** [comment]
So here's what's left, in my head:

- [ ] Consider cherry-picking on 1.13.x
- [ ] Add docs about storage driver recommendations (which page?)

What am I missing from the above?

**Comment by efx** [comment]
@ilianaw i would find it helpful to have some documentation around the correct SElinux settings or if that is possible on latest AL AMI. 
apologies for hijacking, but what command do I issue for actually building the rpm? `make rpm` within container based off of `dockercore/builder-rmp amazonlinux-latest`?

**Comment by ilianaw** [comment]
@efx SELinux is disabled by default on AL AMI, and we don't have any documentation or recommendations around SELinux settings to use in conjunction with Docker.

For me, I ran `make rpm` from the root of the git repo (I believe `docker` needs to be runnable by your user without sudo, and Docker-in-Docker needs to work). It generates all the RPMs, but it does them in alphabetical order :)

**Comment by efx** [comment]
@ilianaw thanks a ton! this saved me a number of hours getting docker up and running. Good to know SELinux support isn't there too.

**Comment by thaJeztah** [comment]
You can use the `DOCKER_BUILD_PKGS` variable to specify which RPM's to build. For example;

```
make BIND_DIR=. DOCKER_BUILD_PKGS="amazonlinux-latest" rpm
```

**Pull request merged**